README
Net-API-Gett

Gett (http://ge.tt) is a clutter-free file sharing service that allows its users to 
share up to 2 GB of files for free.  They recently implemented a REST API; this is a 
binding for the API. See http://ge.tt/developers for full details and how to get an
API key.

Version: 1.04

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::API::Gett

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-API-Gett

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Net-API-Gett

    CPAN Ratings
        http://cpanratings.perl.org/d/Net-API-Gett

    Search CPAN
        https://metacpan.org/module/Net::API::Gett

    Git Hub
        https://github.com/mrallen1/Net-API-Gett


LICENSE AND COPYRIGHT

Copyright (C) 2011 Mark Allen

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Net-API-Gett

1.05    2012-02-28
        @keedi fixed a failed test (due to internal Gett URL format change)

1.04    2012-02-01
        Reworked a patch from @keedi to chunk file uploads dynamically. This cuts 
        RAM usage substantially.

1.03    2012-01-10
        Patch from @keedi to fix the getturl attribute on file objects.
        Set MooX::Types::MooseLike to be version 0.02 (unreleased version 0.03 breaks API)

1.02    2012-01-02
        Fix for RT #73671 - Thanks to Alexander Ost for the report and patch

1.01    2012-01-01
        Fix test dependency failure (from CPAN testers) D'oh

1.00    2011-12-31
        Refactor to use better object encapsulation.

0.02    2011-12-23
        Integrate patch from keedi to allow object initialization 
        with one of:
            * access token
            * refresh token, or,
            * api key, email and password

        Document change
        Make build dependent on Moo 0.009013 which fixes CPAN test failures.

0.01    2011-12-02
        Initial release



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Net::API::Gett',
    license             => 'perl',
    dist_author         => q{Mark Allen <mrallen1@yahoo.com>},
    dist_version_from   => 'lib/Net/API/Gett.pm',
    configure_requires => {
        'Module::Build' => 0.38,
    },
    build_requires => {
    },
    test_requires => {
        'File::Temp' => 0,
        'Test::More'  => 0,
    },
    requires => {
        'perl' => '5.010',
        'LWP::Protocol::https' => 0,
        'Moo'         => '0.009013',
        'Sub::Quote'  => 0,
        'LWP'         => 0,
        'JSON'        => 0,
        'File::Slurp' => 0,
        'Array::Iterator' => 0,
        'MooX::Types::MooseLike' => 0,
    },
    add_to_cleanup      => [ 'Net-API-Gett-*' ],
    create_makefile_pl => 'traditional',
    meta_merge => {
        resources => {
            repository => 'https://github.com/mrallen1/Net-API-Gett.git',
            KeyWords => ["file sharing", "gett", "REST API client"],
        },
    },
);

$builder->create_build_script();


lib/Net/API/Gett/File.pm
package Net::API::Gett::File;

=head1 NAME

Net::API::Gett::File - Gett file object

=cut

use Moo;
use Carp qw(croak);
use MooX::Types::MooseLike qw(Int Str);

our $VERSION = '1.05';

=head1 PURPOSE

Encapsulate Gett files.  You normally shouldn't instantiate this class on 
its own, as the library will create and return this object as appropriate.

=head1 ATTRIBUTES

These are read only attributes unless otherwise noted. 

=over 

=item filename

Scalar string.

=item fileid

Scalar string.

=item downloads

Scalar integer. The number of times this particular file has been downloaded

=item readystate

Scalar string. Signifies the state a particular file is in. See the 
L<Gett developer docs|http://ge.tt/developers> for more information.

=item getturl

Scalar string. The URL to use in a browser to access a file.

=item download

Scalar string. The URL to use to get the file contents.

=item size

Scalar integer. The size in bytes of this file.

=item created

Scalar integer. The Unix epoch time when this file was created in Gett. This value is suitable
for use in C<localtime()>.

=item sharename

Scalar string.  The share in which this file lives inside.

=item put_upload_url

Scalar string.  The url to use to upload the contents of this file using the PUT method. (This
attribute is only populated during certain times.)

=item post_upload_url

Scalar string. This url to use to upload the contents of this file using the POST method. (This
attribute is only populated during certain times.)

=item chunk_size

Scalar integer. This is the chunk size to use for file uploads. It defaults to 
1,048,576 bytes (1 MB).  This attribute is read-only.

=back

=cut

has 'filename' => (
    is => 'ro',
    isa => Str,
);

has 'fileid' => (
    is => 'ro',
    isa => Str,
);

has 'downloads' => (
    is => 'ro',
    isa => Int,
);

has 'readystate' => (
    is => 'ro',
    isa => Str,
);

has 'getturl' => (
    is => 'ro',
    isa => Str,
);

has 'download' => (
    is => 'ro',
    isa => Str,
);

has 'size' => (
    is => 'ro',
    isa => Int,
);

has 'created' => (
    is => 'ro',
    isa => Int,
);

has 'sharename' => (
    is => 'rw',
    isa => Str,
);

has 'put_upload_url' => (
    is => 'ro',
    isa => Str,
);

has 'post_upload_url' => (
    is => 'ro',
    isa => Str,
);

has 'chunk_size' => (
    is => 'rw',
    isa => Int,
    default => sub { 1_048_576; },
);

=over

=item user

L<Net::API::Gett::User> object. C<has_user()> predicate.

=back

=cut

has 'user' => (
    is => 'rw',
    predicate => 'has_user',
    isa => sub { die "$_[0] is not Net::API::Gett::User" unless ref($_[0]) =~ /User/ },
    lazy => 1,
);

=over

=item request

L<Net::API::Gett::Request> object.

=back

=cut

has 'request' => (
    is => 'rw',
    isa => sub { die "$_[0] is not Net::API::Gett::Request" unless ref($_[0]) =~ /Request/ },
    default => sub { Net::API::Gett::Request->new() },
    lazy => 1,
);

=head1 METHODS

=over

=item send_file()

This method actually uploads the file to the Gett service. This method is normally invoked by the
C<upload_file()> method, but it might be useful in combination with C<get_upload_url()>. It takes 
the following parameters:

=over

=item * a Gett put upload url

=item * data

a scalar representing the file contents which can be one of: a buffer, an L<IO::Handle> object, or a 
file pathname.

=item * encoding

an encoding scheme. By default, it uses C<:raw>.

=item * chunk_size

The maximum chunksize to load into to memory at one time.  If the file to transmit is larger than
this size, it will be dynamically streamed.

=back

Returns a true value on success.

=back

=cut

sub send_file {
    my $self = shift;
    my $url = shift;
    my $contents = shift;
    my $encoding = shift || ":raw";
    my $chunk_size = shift || $self->chunk_size || 1_048_576; # 1024 * 1024 = 1 MB

    my $fh;
    my $length;
    if ( ! ref($contents) ) {
        # $contents is scalar
        if ( ! -e $contents ) {
            # $contents doesn't point to a valid filename, 
            # assume it's a buffer

            $contents .= "";
            # Make sure this data is stringified.
            open($fh, "<", \$contents) or croak "Couldn't open a filehandle on the content buffer\n";
            binmode($fh, $encoding);
            $length = length($contents);
        }
        else {
            open($fh, "<", $contents) or croak "Couldn't open a filehandle on $contents: $!";
            binmode($fh, $encoding);
            $length = -s $contents;
        }
    }
    else {
       $fh = $contents if ref($contents) =~ /IO/;
       $length = -s $fh;
    }

    return 0 unless $fh;

    my $response = $self->request->put($url, $fh, $chunk_size, $length);

    if ( $response ) {
        return 1;
    }
    else {
        return undef;
    }
}

=over

=item get_upload_url()

This method returns a scalar PUT upload URL for the specified sharename/fileid parameters. 
Potentially useful in combination with C<send_file()>.

=back

=cut

sub get_upload_url {
    my $self = shift;
    croak "Cannot get_upload_url() without a Net::API::Gett::User object." unless $self->has_user;

    my $sharename = $self->sharename;
    my $fileid = $self->fileid;

    $self->user->login unless $self->user->has_access_token;

    my $endpoint = "/files/$sharename/$fileid/upload?accesstoken=".$self->user->access_token;

    my $response = $self->request->get($endpoint);

    if ( $response && exists $response->{'puturl'} ) {
        return $response->{'puturl'};
    }
    else {
        croak "Could not get a PUT url from $endpoint";
    }
}

=over

=item destroy()

This method destroys the file represented by the object. Returns a true value on success.

=back

=cut

sub destroy {
    my $self = shift;
    croak "Cannot destroy() without a Net::API::Gett::User object." unless $self->has_user;

    my $sharename = $self->sharename;
    my $fileid = $self->fileid;

    $self->user->login unless $self->user->has_access_token;

    my $endpoint = "/files/$sharename/$fileid/destroy?accesstoken=".$self->access_token;

    my $response = $self->request->post($endpoint);

    if ( $response ) {
        return 1;
    }
    else {
        return undef;
    }
}
        
sub _file_contents {
    my $self = shift;
    my $endpoint = $self->request->base_url . shift;

    my $response = $self->request->ua->get($endpoint);

    if ( $response->is_success ) {
        return $response->content();
    }
    else {
        croak "$endpoint said " . $response->status_line;
    }
}

=over

=item contents()

This method retrieves the contents of a this file in the Gett service.  You are responsible for 
outputting the file (if desired) with any appropriate encoding. Does not require an access token.

=back

=cut

sub contents {
    my $self = shift;
    my $sharename = $self->sharename;
    my $fileid = $self->fileid;

    return $self->_file_contents("/files/$sharename/$fileid/blob");
}

=over

=item thumbnail()

This method returns a thumbnail if the file in Gett is an image. Does not require an access token, but
is really only meaningful if the data is a valid image format file.

=back

=cut

sub thumbnail {
    my $self = shift;
    my $sharename = $self->sharename;
    my $fileid = $self->fileid;

    return $self->_file_contents("/files/$sharename/$fileid/blob/thumb");
}

=head1 SEE ALSO

L<Net::API::Gett>

=cut

1;


lib/Net/API/Gett/User.pm
package Net::API::Gett::User;

use Moo;
use Carp qw(croak);
use MooX::Types::MooseLike qw(Str Int);

use Net::API::Gett::Request;

our $VERSION = '1.05';

=head1 NAME

Net::API::Gett::User - Gett User object

=head1 PURPOSE

This class encapsulates Gett service user functions. You normally shouldn't instanstiate 
this class on its own as the library will create and return this object when appropriate.

=head1 ATTRIBUTES

Here are the attributes of this class.  They are read only.

=over 

=item api_key

Scalar string. C<has_api_key> predicate.

=back

=cut

has 'api_key' => ( 
    is  => 'ro',
    predicate => 'has_api_key',
    isa => Str,
);

=over 

=item email

Scalar string. C<has_email> predicate.

=back

=cut

has 'email' => (
    is  => 'ro',
    predicate => 'has_email',
    isa => sub { die "$_[0] is not email" unless $_[0] =~ /.+@.+/ },
);

=over

=item password

Scalar string. C<has_password> predicate.

=back

=cut

has 'password' => (
    is  => 'ro',
    predicate => 'has_password',
    isa => Str,
);

=over

=item access_token

Scalar string. Populated by C<login> call. C<has_access_token()> predicate.

=back 

=cut

has 'access_token' => (
    is        => 'rw',
    writer    => '_set_access_token',
    predicate => 'has_access_token',
    isa => Str,
);

=over

=item access_token_expiration

Scalar integer. Unix epoch seconds until an access token is no longer valid which is 
currently 24 hours (86400 seconds) from token acquisition. 
This value is suitable for use in a call to C<localtime()>.

=back

=cut

has 'access_token_expiration' => (
    is        => 'rw',
    writer    => '_set_access_token_expiration',
    isa => Int,
);

=over

=item refresh_token

Scalar string. Populated by C<login> call.  Can be used to generate a new valid
access token without reusing an email/password login method.  C<has_refresh_token()> 
predicate.

=back

=cut

has 'refresh_token' => (
    is        => 'rw',
    writer    => '_set_refresh_token',
    predicate => 'has_refresh_token',
    isa => Str,
);

=over

=item request

This is a L<Net::API::Gett::Request> object. Defaults to a new instance of that class.

=back

=cut

has 'request' => (
    is => 'rw',
    isa => sub { die "$_[0] is not Net::API::Gett::Request" unless ref($_[0]) =~ /Request/ },
    default => sub { Net::API::Gett::Request->new() },
    lazy => 1,
);

=over

=item userid

Scalar string.

=item fullname

Scalar string.

=item email

Scalar string.

=item storage_used

Scalar integer. In bytes.

=item storage_limit

Scalar integer. In bytes.

=back


=cut

has 'userid' => (
    is => 'rw',
    writer => '_set_userid',
    isa => Str
);

has 'fullname' => (
    is => 'rw',
    writer => '_set_fullname',
    isa => Str,
);

has 'storage_used' => (
    is => 'rw',
    isa => Int,
    writer => '_set_storage_used',
);

has 'storage_limit' => (
    is => 'rw',
    isa => Int,
    writer => '_set_storage_limit',
);

=head1 METHODS 

=over

=item login()

This method populates the C<access_token>, C<refresh_token> and C<user> attributes.  It usually
doesn't need to be explicitly called since methods which require an access token will automatically
(and lazily) attempt to log in to the API and get one.

Returns a perl hash representation of the JSON output for L<https://open.ge.tt/1/users/login>.

=back

=cut

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;

    my %params = @_;

    unless (
           $params{refresh_token}
        || $params{access_token}
        || ( $params{api_key} && $params{email} && $params{password} ) )
    {
        die(
            "api_key, email and password are needed to create ",
            "Net::API::Gett::User object. Or you can use refresh_token ",
            "or access_token rather than api_key, email and password.\n",
        );
    }

    return $class->$orig(@_);
};

sub login {
    my $self = shift;

    my %hr;
    if ( $self->has_refresh_token ) {
        $hr{'refreshtoken'} = $self->refresh_token;
    }
    elsif ( $self->has_api_key && $self->has_email && $self->has_password ) {
        @hr{'apikey', 'email', 'password'} = (
            $self->api_key,
            $self->email,
            $self->password);
    }
    else {
        croak "I need either an api_key, email, and password or a refresh token to login";
    }

    my $response = $self->request->post('/users/login', \%hr);

    # $response is a hashref
    # see https://open.ge.tt/1/doc/rest#users/login for response keys

    if ( $response ) {
        $self->_set_access_token( $response->{'accesstoken'} );
        $self->_set_access_token_expiration( time + $response->{'expires'} );
        $self->_set_refresh_token( $response->{'refreshtoken'} );
        $self->_set_attrs( $response->{'user'} );
        return $self;
    }
    else {
        croak("No response from user->login");
    }
}

=over

=item refresh()

Refreshes user data.

=back

=cut

sub refresh {
    my $self = shift;

    $self->login unless $self->has_access_token;

    my $endpoint = "/users/me?accesstoken=" . $self->access_token;

    my $response = $self->request->get($endpoint);

    if ( $response ) {
        $self->_set_attrs($response);
        return $self;
    }
    else {
        croak("No response from user->refresh");
    }
}

sub _set_attrs {
    my $self = shift;
    my $uref = shift; # hashref https://open.ge.tt/1/doc/rest#users/me

    return undef unless ref($uref) eq "HASH";

    $self->_set_userid($uref->{'userid'});
    $self->_set_fullname($uref->{'fullname'});
    $self->_set_storage_used($uref->{'storage'}->{'used'});
    $self->_set_storage_limit($uref->{'storage'}->{'limit'}),
}

=head1 SEE ALSO

L<Net::API::Gett>

=cut

1;



lib/Net/API/Gett/Share.pm
package Net::API::Gett::Share;

=head1 NAME

Net::API::Gett::Share - Gett share object

=cut

use Moo;
use Carp qw(croak);
use Array::Iterator;
use MooX::Types::MooseLike qw(Int Str);

our $VERSION = '1.05';

=head1 PURPOSE

Encapsulate Gett shares.  You normally shouldn't instantiate this class on its own, as the
library will create and return this object as appropriate.

=head1 ATTRIBUTES

=over

=item user

L<Net::API::Gett::User> object. C<has_user()> predicate.

=back

=cut

has 'user' => (
    is => 'rw',
    predicate => 'has_user',
    isa => sub { die "$_[0] is not Net::API::Gett::User" unless ref($_[0]) =~ /User/ },
    lazy => 1,
);

=over

=item request

L<Net::API::Gett::Request> object.

=back

=cut

has 'request' => (
    is => 'rw',
    isa => sub { die "$_[0] is not Net::API::Gett::Request" unless ref($_[0]) =~ /Request/ },
    default => sub { Net::API::Gett::Request->new() },
    lazy => 1,
);

=over 

=item sharename

Scalar string. Read only.

=item title

Scalar string.

=item created 

Scalar integer. Read only. This value is in Unix epoch seconds, suitable for use in a call to C<localtime()>.

=item getturl

Scalar string. The URL to use in a browser to access a share.

=item files

This attribute holds any L<Net::API::Gett:File> objects linked to a particular 
share instance. It returns a list of L<Net::API::Gett:File> objects if 
there are any, otherwise returns an empty list. See also 
C<get_file_iterator()> below.

=back

=cut

has 'sharename' => (
    is => 'ro',
    isa => Str,
);

has 'title' => (
    is => 'rw',
);

has 'created' => (
    is => 'ro',
    isa => Int,
);

has 'getturl' => (
    is => 'ro',
    isa => Str,
);

sub files {
    my $self = shift;

    return () unless exists $self->{'files'};

    return @{ $self->{'files'} };
}

=head1 METHODS

=over

=item add_file()

This method stores a new L<Net::API::Gett::File> object in the share object.
It returns undef if the value passed is not an L<Net::API::Gett::File> object.

=back

=cut

sub add_file {
    my $self = shift;
    my $file = shift;

    return undef unless ref($file) =~ /File/;

    $file->sharename($self->sharename) unless $file->sharename;

    $file->user($self->user) if $self->has_user;

    push @{ $self->{'files'} }, $file;
}

=over

=item file_iterator()

This method returns a L<Array::Iterator> object on the files in this share. For full details 
about L<Array::Iterator> please read its documentation. It supports standard iterator
methods such as:

=over

=item * next()

=item * has_next()

=item * get_next()

=item * peek()

=back

Example code using peek:

  my $share = $gett->get_share("4ddsfds");
  my $file_iter = $share->file_iterator();

  while ( $file_iter->has_next ) {
    if ( $file_iter->peek->size > 1_048_576 ) {
      # Skip big file
      warn $file_iter->peek->filename . " is too large, skipping\n";
      $file_iter->next();
    } 
    else {
      my $file = $file_iter->next();
      printf "name: %s\tsize: %d\n", $file->filename, $file->size;
    }
  }

Example code using get_next:

  for my $file ( $file_iter->get_next() ) {
    say "name: " . $file->filename . "\tsize: " . $file->size;
  }

=back

This method returns undef if there are no files associated with a share.

=cut

sub file_iterator {
    my $self = shift;

    return undef unless exists $self->{'files'};

    return Array::Iterator->new($self->{'files'});
}

=over

=item update()

This method updates share attributes.  At present, only the share title can be changed (or deleted), 
so pass in a string to set a new title for a specific share.

Calling this method with an empty parameter list or explicitly passing C<undef> 
will B<delete> any title currently set on the share.

Returns a L<Net::API::Gett:Share> object with updated values.

=back

=cut
        
sub update {
    my $self = shift;

    croak "Cannot call update without a Net::API::Gett::User object" unless $self->has_user;

    my $title = shift;
    my $name = $self->sharename;

    $self->user->login unless $self->user->has_access_token;

    my $response = $self->request->post("/shares/$name/update?accesstoken=".$self->user->access_token, 
            { title => $title } );

    if ( $response ) {
        $self->title( $response->{'title'} );
        return $self;
    }
    else {
        return undef;
    }
}

=over

=item destroy()

This method destroys the share and all of the share's files.  Returns a true boolean
on success.

=back

=cut

sub destroy {
    my $self = shift;

    croak "Cannot call destroy without a Net::API::Gett::User object" unless $self->has_user;

    my $name = $self->sharename;

    $self->user->login unless $self->user->has_access_token;

    my $response = $self->request->post("/shares/$name/destroy?accesstoken=".$self->user->access_token);

    if ( $response ) {
        return 1;
    }
    else {
        return undef;
    }
}

=head1 SEE ALSO

L<Net::API::Gett>, L<Array::Iterator>

=cut

1;


lib/Net/API/Gett.pm
package Net::API::Gett;

use strict;
use warnings;

use 5.010;

use Moo;
use Carp qw(croak);
use Scalar::Util qw(looks_like_number);

use Net::API::Gett::User;
use Net::API::Gett::Share;
use Net::API::Gett::File;
use Net::API::Gett::Request;

BEGIN {
    require LWP::Protocol::https or die "This module requires HTTPS, please install LWP::Protocol::https\n";
}

=head1 NAME

Net::API::Gett - Perl bindings for Ge.tt API

=head1 VERSION

Version 1.05

=cut

our $VERSION = '1.05';

=head1 SYNOPSIS

    use 5.010;
    use Net::API::Gett;

    # Get API Key from http://ge.tt/developers

    my $gett = Net::API::Gett->new( 
        api_key      => 'GettAPIKey',
        email        => 'me@example.com',
        password     => 'mysecret',
    );

    my $file_obj = $gett->upload_file( 
        filename => "ossm.txt",
        contents => "/some/path/example.txt",
           title => "My Awesome File", 
        encoding => ":encoding(UTF-8)" 
    );

    say "File has been shared at " . $file_obj->getturl;

    # Download contents
    my $file_contents = $file_obj->contents();

    open my $fh, ">:encoding(UTF-8)", "/some/path/example-copy.txt" 
        or die $!;
    print $fh $file_contents;
    close $fh;

    # clean up share and file(s)
    my $share = $gett->get_share($file->sharename);
    $share->destroy();

=head1 ABOUT

L<Gett|http://ge.tt> is a clutter-free file sharing service that allows its users to 
share up to 2 GB of files for free.  They recently implemented a REST API; this is a 
binding for the API. See L<http://ge.tt/developers> for full details and how to get an
API key.

=head1 CHANGES FROM PREVIOUS VERSION

This library is more encapsulated. Share functions which act on shares are in the L<Net::API::Gett::Share> 
object namespace, and likewise with Ge.tt L<files|Net::API::Gett:File>. Future versions of this library 
will modify the L<Request|Net::API::Gett::Request> and L<User|Net::API::Gett::User> objects to be 
L<roles|Moo::Role> rather than objects.

=cut

sub BUILD {
    my $self = shift;
    my $args = shift;

    unless ( $self->has_user ) {
        if ( $args->{refresh_token} ||
             $args->{access_token}  ||
             ( $args->{api_key} && $args->{email} && $args->{password} ) ) {
                $self->user( Net::API::Gett::User->new(%{$args}) );
        }
    }
}

=head1 ATTRIBUTES

=over

=item user

L<Net::API::Gett::User> object. C<has_user()> predicate.

=back

=cut

has 'user' => (
    is => 'rw',
    lazy => 1,
    predicate => 'has_user',
    isa => sub { die "$_[0] is not Net::API::Gett::User" unless ref($_[0]) =~ /User/ },
);

=over

=item request

L<Net::API::Gett::Request> object.

=back

=cut

has 'request' => (
    is => 'rw',
    isa => sub { die "$_[0] is not Net::API::Gett::Request" unless ref($_[0]) =~ /Request/ },
    default => sub { Net::API::Gett::Request->new() },
    lazy => 1,
);

=head1 METHODS

Unless otherwise noted, these methods die if an error occurs or if they get a response from the API
which is not successful. If you need to handle errors more gracefully, use L<Try::Tiny> to catch fatal 
errors.

=over

=item new()

Constructs a new object.  Optionally accepts:

=over

=item * A Ge.tt API key, email, and password, or,

=item * A Ge.tt refresh token, or,

=item * A Ge.tt access token

=back

If any of these parameters are passed, they will be proxied into the L<Net::API::Gett::User> object which
will then permit you to make authenticated API calls.  Without an access token (or the means to acquire one)
only non-authenticated calls are allowed; they are: C<get_share()>, C<get_file()>, 
C<$file-E<gt>thumbnail()> and C<$file-E<gt>contents()>.

=back

=head2 Share functions

All of these functions cache L<Net::API::Gett::Share> objects. Retrieve objects from the 
cache using the C<shares> method.  Use the C<get_share> method to update a cache entry if it
is stale.

=over

=item get_shares()

Retrieves B<all> share information for the given user.  Takes optional scalar integers 
C<offset> and C<limit> parameters, respectively. 

Returns an unordered list of L<Net::API::Gett::Share> objects. 

=back

=cut

sub get_shares {
    my $self = shift;
    croak "Can't call get_shares() without Net::API::Gett::User object" unless $self->has_user;

    my $offset = shift;
    my $limit = shift;

    $self->user->login unless $self->user->has_access_token;

    my $endpoint = "/shares?accesstoken=" . $self->user->access_token;

    if ( defined $offset && looks_like_number $offset ) {
        $endpoint .= "&skip=$offset";
    }

    if ( defined $limit && looks_like_number $limit ) {
        $endpoint .= "&limit=$limit";
    }

    my $response = $self->request->get($endpoint);

    if ( $response ) {
        foreach my $share_href ( @{ $response } ) {
            next unless $share_href;
            $self->add_share(
                $self->_build_share($share_href)
            );
        }
        return $self->shares;
    }
    else {
        return undef;
    }
}

=over

=item get_share()

Retrieves (and/or refreshes cached) information about a specific single share. 
Requires a C<sharename> parameter. Does not require an access token to call.

Returns a L<Net::API::Gett::Share> object.

=back

=cut

sub get_share {
    my $self = shift;
    my $sharename = shift;

    return undef unless $sharename =~ /\w+/;

    my $response = $self->request->get("/shares/$sharename");

    if ( $response ) {
        my $share = $self->_build_share($response);
        $self->add_share($share);
        return $share;
    }
    else {
        return undef;
    }
}

=over

=item create_share()

This method creates a new share instance to hold files. Takes an optional string scalar
parameter which sets the share's title attribute.

Returns the new share as a L<Net::API::Gett::Share> object.

=back

=cut

sub create_share {
    my $self = shift;
    croak "Can't call create_share() without Net::API::Gett::User object" unless $self->has_user;

    my $title = shift;

    $self->user->login unless $self->user->has_access_token;

    my @args = ("/shares/create?accesstoken=".$self->user->access_token);
    if ( $title ) {
        push @args, { title => $title };
    }
    my $response = $self->request->post(@args);

    if ( $response ) {
        my $share = $self->_build_share($response);
        $self->add_share($share);
        return $share;
    }
    else {
        return undef;
    }
}

=head2 File functions

=over

=item get_file()

Returns a L<Net::API::Gett::File> object given a C<sharename> and a C<fileid>.
Does not require an access token to call.

=back

=cut

sub get_file {
    my $self = shift;
    my $sharename = shift;
    my $fileid = shift;

    my $response = $self->request->get("/files/$sharename/$fileid");

    if ( $response ) {
        return $self->_build_file($response);
    }
    else {
        return undef;
    }
}

=over

=item upload_file()

This method uploads a file to Gett. The following key/value pairs are valid:

=over

=item * filename (B<required>) 
    
What to call the uploaded file when it's inside of the Gett service.

=item * sharename (optional) 
    
Where to store the uploaded file. If not specified, a new share will be automatically created.

=item * title (optional) 
    
If specified, this value is used when creating a new share to hold the file. It will not change
the title of an existing share. See the C<update()> method on the share object to do that.

=item * content (optional)

A synonym for C<contents>. (Yes, I've typo'd this too many times.) Anything in C<contents> has 
precedent, if they're both specified.

=item * contents (optional) 

A representation of the file's contents.  This can be one of:

=over

=item * A buffer (See note below)

=item * An L<IO::Handle> object

=item * A FILEGLOB

=item * A pathname to a file to be read

=back

If not specified, the C<filename> parameter is treated as a pathname. This attempts to be DWIM, 
in the sense that if C<contents> contains a value which is not a valid filename, it treats 
C<contents> as a buffer and uploads that data.

=item * encoding

An encoding scheme for the file content. By default it uses C<:raw>. See C<perldoc -f binmode> 
for more information about encodings.

=item * chunk_size

The chunk_size in bytes to use when uploading a file.  Defaults to 1 MB.

=back

Returns a L<Net::API::Gett:File> object representing the uploaded file.

=back

=cut

sub upload_file {
    my $self = shift;
    my $opts = { @_ };

    return undef unless ref($opts) eq "HASH";

    my $sharename = $opts->{'sharename'};

    if ( not $sharename ) {
        my $share = $self->create_share($opts->{'title'});
        $sharename = $share->sharename;
    }

    $self->user->login unless $self->user->has_access_token;

    my $endpoint = "/files/$sharename/create?accesstoken=".$self->user->access_token;
    
    my $filename = $opts->{'filename'};

    my $response = $self->request->post($endpoint, { filename => $filename });

    # typo proof this - yeah I've been bitten by this!
    unless ( exists $opts->{'contents'} ) {
            if ( exists $opts->{'content'} ) {
            $opts->{'contents'} = delete $opts->{'content'};
        }
        else {
            $opts->{'contents'} = $filename 
        }
    }

    if ( $response ) {
        my $file = $self->_build_file($response);
        if ( $file->readystate eq "remote" ) {
            my $put_upload_url = $file->put_upload_url;
            croak "Didn't get put upload URL from $endpoint" unless $put_upload_url;
            if ( $file->send_file($put_upload_url, $opts->{'contents'}, 
                        $opts->{'encoding'}, $opts->{'chunk_size'}) ) {
                return $file;
            }
            else {
                croak "There was an error reading data from " . $opts->{'contents'};
            }
        }
        else {
            croak "$endpoint doesn't have right readystate";
        }
    }
    else {
        return undef;
    }
}

sub _build_share {
    my $self = shift;
    my $share_href = shift;

    my $share = Net::API::Gett::Share->new(
        sharename => $share_href->{'sharename'},
        created => $share_href->{'created'},
        title => $share_href->{'title'},
        getturl => $share_href->{'getturl'},
    );
    foreach my $file_href ( @{ $share_href->{'files'} } ) {
        next unless $file_href;
        my $file = $self->_build_file($file_href);
        $share->add_file($file);
    }
    $share->user($self->user) if $self->has_user;

    return $share;
}

sub _build_file {
    my $self = shift;
    my $file_href = shift;

    # filter out undefined attributes
    my @attrs = grep { defined $file_href->{$_} } 
        qw(filename size created fileid downloads readystate getturl download sharename);
    my @params = map { $_ => $file_href->{$_} } @attrs;

    if ( exists $file_href->{'upload'} ) {
        push @params, 'put_upload_url' => $file_href->{'upload'}->{'puturl'};
        push @params, 'post_upload_url' => $file_href->{'upload'}->{'posturl'};
    }

    my $file = Net::API::Gett::File->new( @params );
    $file->user($self->user) if $self->has_user;
    
    return $file;
}

=over

=item add_share()

This method populates/updates the L<Net::API::Gett:Share> object local cache.

It returns undef if the passed value isn't a L<Net::API::Gett::Share> object.

=back

=cut

sub add_share {
    my $self = shift;
    my $share = shift;

    return undef unless ref($share) =~ /Share/;

    my $sharename = $share->sharename();

    $self->{'shares'}->{$sharename} = $share;
}

=over

=item shares()

This method retrieves one or more cached L<Net::API::Gett::Share> objects. Objects are
requested by sharename.  If no parameter list is specified, B<all> cached objects are 
returned in an unordered list. (The list will B<not> be in the order shares were added
to the cache.)

If no objects are cached, this method returns an empty list.

=back

=cut

sub shares {
    my $self = shift;

    if ( @_ ) {
        return map { $self->{'shares'}->{$_} } @_;
    }

    return () unless exists $self->{'shares'};

    return values %{ $self->{'shares'} };
}

=head1 AUTHOR

Mark Allen, C<< <mrallen1 at yahoo dot com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-api-gett at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-API-Gett>.  I will 
be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::API::Gett

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-API-Gett>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-API-Gett>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-API-Gett>

=item * MetaCPAN

L<https://metacpan.org/module/Net::API::Gett/>

=item * GitHub

L<https://github.com/mrallen1/Net-API-Gett>

=back

=head1 SEE ALSO

L<Gett API documentation|http://ge.tt/developers>

=head1 CONTRIBUTORS

Thanks to the following for patches: 

=over

=item

Keedi Kim (L<https://github.com/keedi>)

=item

Alexander Ost

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Mark Allen.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of Net::API::Gett
